### DISCUSSION TITLE
This Process Step has already occurred.

### DISCUSSION THREAD

**Discussion by alexanderkrs** [flowsforapex/apex-flowsforapex#706] [Q&A] [answered]
Hi everyone,
I would like some help in a process that I have worked out. Through a query, I get: 
sbfl_prcs_id,
sbfl_id 
sbfl_step_key

With this I use 
A procedure flow_api_pkg.flow_complete_step 
Due to the need to complete an activity in 2 different instances, the data required to fill it out comes as follows:
sbfl_prcs_id = 290,291
sbfl_id = 319,320
sbfl_step_key = WlKgrdaHis,cyPPQDbunq 

When I run the procedure with this data, the error appears: 
"This Process Step has already occurred. (Incorrect step key WlKgrdaHis supplied while expecting step key cyPPQDbunq)." 

I would like some help to solve this problem.

**Comment by rallen2010** [comment]
You can only issue the flow_complete_step command once for each process step - because that command steps the process forward onto the next step.  If you try to step forward from a step more than once, the second (and subsequent) attempts will fail with the message that you are getting.

Your note says that you need to step two separate process instances forward from your program.  It looks as if you are supplying the wrong step key to the wrong process - so you'd need to do:

```sql
flow_api_pkg.flow_complete_step (process_id => 290, subflow_id => 319, step_key => 'WlKgrdaHis');
flow_api_pkg.flow_complete_step (process_id => 291, subflow_id => 320, step_key => 'cyPPQDbunq');
```
Note that this will cause a commit to happen after each flow_complete_step command - so this will occur as two separate transactions.  You can't currently make this all part of a single transaction - so you will have to handle any issues from that.

We are developing a feature at present for process collaboration using MessageFlow - which allows one process instance to start another process, to subscribe for and wait for a message in another instance, to send a message to another instance, to create interrupting and non-interrupting boundary events on tasks, subProcesses and callActivities, and to throw a message from an endEvent.

**Reply by alexanderkrs** [reply]
I was able to understand that the problem could be this, so I made a code to get the items from the correct instance for this I used it like this:

```
Select 
LISTAGG(distinct sbfl_prcs_id,','),
LISTAGG(distinct sbfl_id,','),
LISTAG(distinct sbfl_step_key,',')
into
:P40_PRCS_ID_SELECTED,
:P40_SELECTED_SUBFLOWS,
:P40_SELECTED_STEPKEYS
from MY_TABLE
and SBFL_CURRENT = 'Activity_03ifour' 
```

With that I take the necessary information, and move on to the following code to complete the activities that the current instance is doing:

```
DECLARE
    v_prcs apex_t_varchar2 := apex_string.split(:P 40_PRCS_ID_SELECTED, ',');
    v_subflows apex_t_varchar2 := apex_string.split(:P 40_SELECTED_SUBFLOWS, ',');
    v_step apex_t_varchar2 := apex_string.split(:P 40_SELECTED_STEPKEYS, ',');
BEGIN
    for i in 1..v_prcs.count
        LOOP 
            flow_api_pkg.flow_complete_step(
                p_process_id => v_prcs(i),
                p_subflow_id => v_subflows(i), 
                p_step_key => v_step(i)
            );
        END LOOP; 
EXCEPTION WHEN OTHERS THEN
    DBMS_OUTPUT. PUT_LINE('Flows already updated');
END;
```

This code worked in some situations, however in this current case, it has this problem. 
Maybe with this information you can give me a light, I take the subflow items and step_key from the view flow_subflows_vw as in the examples.

**Reply by rallen2010** [reply, marked as answer]
I suspect the problem is that doing distinct on the selected columns means that the ordering of  items is not maintained - which is consistent with the problem that you are having and the fact that sometimes it works and sometimes it doesn't.  You want to select process_id, subflow_id and matching step key as a set for each process - and use them together.  In processing a distinct and the listagg, your items may well be getting sorted by the database to implement the distinct.  The process id and subflow id will always sort in ascending numerical order, but if the step key gets sorted and stepkey2 sorts before stepkey1, you will get them mixed up ... which looks like what is happening.

**Reply by alexanderkrs** [reply]
Your idea makes sense, I used distinct because of a scenario where I used a subflow and needed only one line of the flow, the most current.
I'm going to test without the distinct and bring the return.

**Reply by alexanderkrs** [reply]
Hello
I would like to inform you that the problem was solved, I built a new query in which I did not need to use DISTINCT, through this I was able to get the necessary information to complete the flow activity. 
Thanks for the help.
